### PR TITLE
Update Automattic-Tracks to 3.0.0 and address breaking changes

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR updates [Automattic-Tracks-Android](https://github.com/Automattic/Automattic-Tracks-Android) to `3.0.0` and addresses the breaking changes by attempting to keep the existing behavior unchanged.

Note that the PR is targeting `deps/main-batch-androidx-compose-kotlin` branch - from https://github.com/wordpress-mobile/WordPress-Android/pull/18364 - because [Tracks library is using Kotlin version `1.8.21` ](https://github.com/Automattic/Automattic-Tracks-Android/blob/trunk/settings.gradle#L4) and `trunk` is at `1.6.10`. Luckily, @ParaskP7 already [did the work to update the Kotlin version to `1.8.21`](https://github.com/wordpress-mobile/WordPress-Android/pull/18364/files#diff-7f825392aa37acd1cee0c2e7b9bb7366ad6eac64f3e6cdd816e156bcb69d30deR2) in https://github.com/wordpress-mobile/WordPress-Android/pull/18364. ❤️ 

As far as I can tell, there are 5 breaking changes from `2.2.0` to `3.0.0`:

1. `CrashLoggingProvider` now requires an `Application` instead of `Context`.
2. `CrashLoggingProvider` now has a new `CoroutineScope` parameter - I used `@Named(APPLICATION_SCOPE)` for this.
3. `CrashLoggingDataProvider` now has a `val user: Flow<CrashLoggingUser?>` property that needs to be overridden instead of `fun userProvider(): CrashLoggingUser?`. I wrapped the current value with `MutableStateFlow` which is what [WCAndroid is using](https://github.com/woocommerce/woocommerce-android/blob/13.5/WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/WCCrashLoggingDataProvider.kt#L45). I originally wrapped it with `flow {}` however when I tested the implementation the user information wasn't available in Sentry. After I made the change in d7b97dfdf7d2e161dbbde04e4b8bbb2686feb0e6, the information showed up as expected.
4. `CrashLoggingDataProvider.applicationContextProvider` now requires a `Flow` - I wrapped the current value with  `MutableStateFlow` for consistency but I think wrapping it with `flow {}` and emitting the single value should work as well.
5. `CrashLoggingDataProvider` now has a `val performanceMonitoringConfig: PerformanceMonitoringConfig` property that needs to be overridden. I've used `PerformanceMonitoringConfig.Disabled` for this to keep the existing behaviour. I'll follow this up with another PR that enables performance monitoring.

Note that I haven't worked with coroutines for a long time and I have never worked with `Flow` before, so I'd appreciate any feedback on what I might have missed or if there is a better way to achieve the changes I tried to make.

**To test:**
1. Throw an exception from `UnifiedAboutActivity.onCreate`
2. Run the app in `release` variant - I don't think it works for `debug` variant, but I am not 100%
3. Navigate to the About page and verify that the app crashes (Tap profile icon on the top right and tap on "About WordPress")
4. Check the latest events in Sentry and verify that the crash is reported correctly - [Here is the crash from my test](https://a8c.sentry.io/share/issue/efb8a23eb68f4d71b8f5eea2ad3e2ab8/).
5. Verify that your user information is showing up correctly in the Sentry event.

## Regression Notes
1. Potential unintended areas of impact
This might possibly break Sentry integration for the project

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I manually crashed the app and verified that the crash is reported to Sentry

3. What automated tests I added (or what prevented me from doing so)
N/A

---

@wzieba @ParaskP7 Since this PR depends on another PR, I am leaving it in draft state, but I was hoping that you could still review it so that I can build on it to add Sentry performance monitoring.

---

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] Talkback.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] Large and small screen sizes. (Tablet and smaller phones)
- [x] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Review by wzieba — APPROVED** [review]
Thank you for dealing with those breaking changes @oguzkocer . Since last release, a few of them accumulated. I hope the documentation I've left in the interface made things easier 👍 .

> Luckily, @ParaskP7 already [did the work to update the Kotlin version to 1.8.21](https://github.com/wordpress-mobile/WordPress-Android/pull/18364/files#diff-7f825392aa37acd1cee0c2e7b9bb7366ad6eac64f3e6cdd816e156bcb69d30deR2) in https://github.com/wordpress-mobile/WordPress-Android/pull/18364. ❤️

👏 🙂

> I originally wrapped it with flow {} however when I tested the implementation the user information wasn't available in Sentry.

I can't see the reason for this, but I believe using the `StateFlow` in this case is a good choice. It's definitely interesting, though, why `flow { emit }` didn't work. Maybe the timing between creating the flow and emitting value?

One thing that concerns me (not a blocker) a little is that before sending events, each time we'll create a new flow for getting a user. I do not know if this is a noticeable performance concern, though. `WCAndroid` handled it via having `crashLoggingUser: MutableStateFlow` property in its `CrashLoggingDataProvider`. But I understand it's not ideal, as it introduces a state.

**Comment by wzieba** [comment]
About testing: just after logging in and going to "About screen", the user in ticket was `id:0` ([link](https://a8c.sentry.io/share/issue/4b26ddd5a10942e39660fa569ce39c44/)). But after trying again 2 times, the correct user was applied ([link](https://a8c.sentry.io/share/issue/4b26ddd5a10942e39660fa569ce39c44/)). 

Probably `accountStore` didn't finish update of `account` field before I went to "About" and crashed the app. Doesn't sound like a blocker to me, leaving just FYI.

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

> Thank you for dealing with those breaking changes @oguzkocer . Since last release, a few of them accumulated. I hope the documentation I've left in the interface made things easier 👍 .

💯 Thank you indeed and thank you @wzieba for taking a look, your input is very valuable here! 🥇 

> One thing that concerns me (not a blocker) a little is that before sending events, each time we'll create a new flow for getting a user. I do not know if this is a noticeable performance concern, though. WCAndroid handled it via having crashLoggingUser: MutableStateFlow property in its CrashLoggingDataProvider. But I understand it's not ideal, as it introduces a state.

I am also thinking about that and whether we should avoid the `get()` pattern here. Looking at what WCAndroid did, I am seeing that their solution looks more closely to the below alternative, which also seems to be working just fine for some time now: 🤔

```Kotlin

...

    override val applicationContextProvider = MutableStateFlow<Map<String, String>>(emptyMap())

...

    override val user = MutableStateFlow(
            accountStore.account?.let { accountModel ->
                CrashLoggingUser(
                    userID = accountModel.userId.toString(),
                    email = accountModel.email,
                    username = accountModel.userName
                )
            })

   override val performanceMonitoringConfig = PerformanceMonitoringConfig.Disabled

...

```

But, doing so, will fail a couple of unit tests (ish) as they would also need to utilize this `reinitialize()` trick, due to the nature of the field having been initialized already, which all it does it to just call `setUp()` again, but naming that re-initialize (I guess for clarity purposes only).

So, for the `should correctly reads and maps user` and `should not provide user if user does not exist` tests, which test `sut.user`, you would need to add the `reinitialize()` call after the arrange and before the act phase of the unit test (see WCAndroid [equivalent tests](https://github.com/woocommerce/woocommerce-android/blob/4f41ee7eac606e71569132ec992d73a93dd23841/WooCommerce/src/test/kotlin/com/woocommerce/android/util/crashlogging/WCCrashLoggingDataProviderTest.kt#L136-L154) for comparison).

The only other test I am seeing needing this `reinitialize()` call would have been the `should provide empty application context`, which test `sut.applicationContextProvider`, but this test doesn't have any arrange anyway, thus not needing that.

For `sut.performanceMonitoringConfig`, it is not tested anywhere, nor on WCAndroid, so I guess can be ignored altogether (having it fixed as `PerformanceMonitoringConfig.Disabled` anyway).

PS: To be honest, I actually think it is a good idea to add the `reinitialize()` call whenever there is an arrange phase on a test, just like WCAndroid is doing, for example adding it on this `should not provide encrypted logging uuid to extras if log file is not available` test (and 4 more tests like that). 🤔

**Comment by oguzkocer** [comment]
@wzieba @ParaskP7 Thank you both for reviewing the PR!

@ParaskP7 I've applied your suggestions in 5366bfe13f72318898457106d2642fd09d83d68f & 467020b40e8456b7b994cfe2a626178b80d05cb5. Let me know if this is what you had in mind.

File: WordPress/src/main/java/org/wordpress/android/util/crashlogging/WPCrashLoggingDataProvider.kt [hunk 1]
```diff
@@ -96,15 +97,17 @@ class WPCrashLoggingDataProvider @Inject constructor(
...
-        return accountStore.account?.let { accountModel ->
-            CrashLoggingUser(
-                userID = accountModel.userId.toString(),
-                email = accountModel.email,
-                username = accountModel.userName
-            )
-        }
-    }
+    override val user = MutableStateFlow(
+            accountStore.account?.let { accountModel ->
```
Comment (wzieba) [inline WordPress/src/main/java/org/wordpress/android/util/crashlogging/WPCrashLoggingDataProvider.kt]: After removing `get()`, if an instance of `WPCrashLoggingDataProvider` will persist between changing users in the app, we'll assign the incorrect user (the old one).

> Looking at what WCAndroid did, I am seeing that their solution looks more closely to the below alternative

In `WCAndroid` we keep logged in user in `crashLoggingUser` property and we observe changes via event bus (see [`onAccountChanged`](https://github.com/woocommerce/woocommerce-android/blob/trunk/WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/WCCrashLoggingDataProvider.kt#L49) method).

edit: the `get()` before worked fine, as every time Sentry requested the user (before sending each event), we were creating a new Flow, with just-requested `AccountStore#account`.

Comment (oguzkocer) [reply WordPress/src/main/java/org/wordpress/android/util/crashlogging/WPCrashLoggingDataProvider.kt]: Right.. Thanks for catching that @wzieba! 🙇‍♂️ 

I've updated the implementation in d00c8d917f356fe24ab946042668adb85af575bb & 361ef69eba9ca1b88594988cae5155fe18705ae6 to bring it inline with the implementation in WCAndroid.

I tested the app after 361ef69eba9ca1b88594988cae5155fe18705ae6 by logging in with one user and then switching to another one before creating a crash and it worked as expected. I verified that before this commit, as you said, the previous user was reported. So, I _think_ we are good 🤞

Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/util/crashlogging/WPCrashLoggingDataProvider.kt]: 👋 @oguzkocer and thanks @wzieba for catching [this](https://github.com/wordpress-mobile/WordPress-Android/pull/18437#discussion_r1197440540)! 🥇 

> @ParaskP7 I've applied your suggestions in https://github.com/wordpress-mobile/WordPress-Android/commit/5366bfe13f72318898457106d2642fd09d83d68f & https://github.com/wordpress-mobile/WordPress-Android/commit/467020b40e8456b7b994cfe2a626178b80d05cb5. Let me know if this is what you had in mind.

Totally! 👍 

> I've updated the implementation in https://github.com/wordpress-mobile/WordPress-Android/commit/d00c8d917f356fe24ab946042668adb85af575bb & https://github.com/wordpress-mobile/WordPress-Android/commit/361ef69eba9ca1b88594988cae5155fe18705ae6 to bring it inline with the implementation in WCAndroid.

Exactly, bringing WPAndroid inline with the WCAndroid implementation (which has been tested), I think, will serve us good here. 💯

**Pull request merged**